File: stripe/src/main/java/com/stripe/android/IssuingCardPinService.java
package com.stripe.android;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.stripe.android.exception.APIConnectionException;
import com.stripe.android.exception.APIException;
import com.stripe.android.exception.AuthenticationException;
import com.stripe.android.exception.CardException;
import com.stripe.android.exception.InvalidRequestException;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import org.json.JSONException;

/**
 * Methods for retrieval / update of a Stripe Issuing card
 */
@SuppressWarnings("WeakerAccess")
public class IssuingCardPinService
        implements EphemeralKeyManager.KeyManagerListener<IssuingCardEphemeralKey> {

    private static final String TAG = IssuingCardPinService.class.getName();
    private static final long KEY_REFRESH_BUFFER_IN_SECONDS = 30L;
    private static final String PIN_RETRIEVE = "PIN_RETRIEVE";
    private static final String PIN_UPDATE = "PIN_UPDATE";
    private static final String ARGUMENT_CARD_ID = "cardId";
    private static final String ARGUMENT_VERIFICATION_ID = "verificationId";
    private static final String ARGUMENT_ONE_TIME_CODE = "userOneTimeCode";
    private static final String ARGUMENT_NEW_PIN = "newPin";

    @NonNull
    private final EphemeralKeyManager<IssuingCardEphemeralKey> mEphemeralKeyManager;
    @NonNull
    private final StripeRepository mStripeRepository;
    @NonNull
    private final OperationIdFactory mOperationIdFactory;
    @NonNull
    private final Map<String, IssuingCardPinRetrievalListener> mRetrievalListeners =
            new HashMap<>();
    @NonNull
    private final Map<String, IssuingCardPinUpdateListener> mUpdateListeners =
            new HashMap<>();

    /**
     * Create a IssuingCardPinService with the provided {@link EphemeralKeyProvider}.
     *
     * @param keyProvider an {@link EphemeralKeyProvider} used to get
     *                    {@link IssuingCardEphemeralKey EphemeralKeys} as needed
     */
    @NonNull
    public static IssuingCardPinService create(
            @NonNull Context context,
            @NonNull EphemeralKeyProvider keyProvider) {
        return new IssuingCardPinService(context, keyProvider, Stripe.getAppInfo());
    }

    private IssuingCardPinService(
            @NonNull Context context,
            @NonNull EphemeralKeyProvider keyProvider,
            @Nullable AppInfo appInfo) {
        this(keyProvider, new StripeApiRepository(context, appInfo), new OperationIdFactory());
    }

    @VisibleForTesting
    IssuingCardPinService(
            @NonNull EphemeralKeyProvider keyProvider,
            @NonNull StripeRepository stripeRepository,
            @NonNull OperationIdFactory operationIdFactory) {
        mOperationIdFactory = operationIdFactory;
        mStripeRepository = stripeRepository;
        mEphemeralKeyManager = new EphemeralKeyManager<>(
                keyProvider,
                this,
                KEY_REFRESH_BUFFER_IN_SECONDS,
                null,
                operationIdFactory,
                new IssuingCardEphemeralKey.Factory(),
                true
        );
    }

    /**
     * Retrieves a PIN for a given card
     *
     * @param cardId          the ID of the card (looks like ic_abcdef1234)
     * @param verificationId  the ID of the verification that was sent to the cardholder
     *                        (typically server-side, through /v1/issuing/verifications)
     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
     * @param listener        a listener for either the PIN, or any error that can occur
     */
    public void retrievePin(
            @NonNull String cardId,
            @NonNull String verificationId,
            @NonNull String userOneTimeCode,
            @NonNull IssuingCardPinRetrievalListener listener) {
        final Map<String, Object> arguments = new HashMap<>();
        arguments.put(ARGUMENT_CARD_ID, cardId);
        arguments.put(ARGUMENT_VERIFICATION_ID, verificationId);
        arguments.put(ARGUMENT_ONE_TIME_CODE, userOneTimeCode);

        final String operationId = mOperationIdFactory.create();
        mRetrievalListeners.put(operationId, listener);
        mEphemeralKeyManager.retrieveEphemeralKey(operationId, PIN_RETRIEVE, arguments);
    }

    /**
     * Retrieves a PIN for a given card
     *
     * @param cardId          the ID of the card (looks like ic_abcdef1234)
     * @param newPin          the new desired PIN
     * @param verificationId  the ID of the verification that was sent to the cardholder
     *                        (typically server-side, through /v1/issuing/verifications)
     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
     * @param listener        a listener for either the PIN, or any error that can occur
     */
    public void updatePin(
            @NonNull String cardId,
            @NonNull String newPin,
            @NonNull String verificationId,
            @NonNull String userOneTimeCode,
            @NonNull IssuingCardPinUpdateListener listener) {
        final Map<String, Object> arguments = new HashMap<>();
        arguments.put(ARGUMENT_CARD_ID, cardId);
        arguments.put(ARGUMENT_NEW_PIN, newPin);
        arguments.put(ARGUMENT_VERIFICATION_ID, verificationId);
        arguments.put(ARGUMENT_ONE_TIME_CODE, userOneTimeCode);

        final String operationId = mOperationIdFactory.create();
        mUpdateListeners.put(operationId, listener);
        mEphemeralKeyManager.retrieveEphemeralKey(operationId, PIN_UPDATE, arguments);
    }

    @Override
    public void onKeyUpdate(@NonNull IssuingCardEphemeralKey ephemeralKey,
                            @Nullable String operationId,
                            @Nullable String action,
                            @Nullable Map<String, Object> arguments) {

        if (PIN_RETRIEVE.equals(action)) {
            IssuingCardPinRetrievalListener listener = mRetrievalListeners.remove(operationId);
            if (listener == null) {
                Log.e(TAG, IssuingCardPinService.class.getName() +
                        " was called without a listener");
                return;
            }
            if (arguments == null) {
                listener.onError(
                        CardPinActionError.UNKNOWN_ERROR,
                        "Arguments were lost during the ephemeral key call, " +
                                "this is not supposed to happen," +
                                " please contact support@stripe.com for assistance.",
                        null);
                return;
            }

            final String cardId =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_CARD_ID));
            final String verificationId =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_VERIFICATION_ID));
            final String userOneTimeCode =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_ONE_TIME_CODE));
            try {
                final String pin = mStripeRepository.retrieveIssuingCardPin(cardId, verificationId,
                        userOneTimeCode, ephemeralKey.getSecret());
                listener.onIssuingCardPinRetrieved(pin);

            } catch (InvalidRequestException e) {
                if ("expired".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_EXPIRED,
                            "The one-time code has expired",
                            null);
                } else if ("incorrect_code".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_INCORRECT,
                            "The one-time code was incorrect",
                            null);
                } else if ("too_many_attempts".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
                            "The verification challenge was attempted too many times",
                            null);
                } else if ("already_redeemed".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_ALREADY_REDEEMED,
                            "The verification challenge was already redeemed",
                            null);
                } else {
                    listener.onError(
                            CardPinActionError.UNKNOWN_ERROR,
                            "The call to retrieve the PIN failed, possibly an error " +
                                    "with the verification. Please check the exception.",
                            e);
                }
            } catch (APIConnectionException |
                    APIException |
                    AuthenticationException |
                    JSONException |
                    CardException e) {
                listener.onError(
                        CardPinActionError.UNKNOWN_ERROR,
                        "An error occurred retrieving the PIN, " +
                                "please check the exception",
                        e);
            }
        }
        if (PIN_UPDATE.equals(action)) {

            IssuingCardPinUpdateListener listener = mUpdateListeners.remove(operationId);
            if (listener == null) {
                Log.e(TAG, IssuingCardPinService.class.getName() +
                        " was called without a listener");
                return;
            }
            if (arguments == null) {
                listener.onError(
                        CardPinActionError.UNKNOWN_ERROR,
                        "Arguments were lost during the ephemeral key call, " +
                                "this is not supposed to happen," +
                                " please contact support@stripe.com for assistance.",
                        null);
                return;
            }

            final String cardId =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_CARD_ID));
            final String newPin =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_NEW_PIN));
            final String verificationId =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_VERIFICATION_ID));
            final String userOneTimeCode =
                    (String) Objects.requireNonNull(arguments.get(ARGUMENT_ONE_TIME_CODE));
            try {
                mStripeRepository.updateIssuingCardPin(cardId, newPin, verificationId,
                        userOneTimeCode, ephemeralKey.getSecret());
                listener.onIssuingCardPinUpdated();
            } catch (InvalidRequestException e) {
                if ("expired".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_EXPIRED,
                            "The one-time code has expired",
                            null);
                } else if ("incorrect_code".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_INCORRECT,
                            "The one-time code was incorrect",
                            null);
                } else if ("too_many_attempts".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
                            "The verification challenge was attempted too many times",
                            null);
                } else if ("already_redeemed".equals(e.getErrorCode())) {
                    listener.onError(
                            CardPinActionError.ONE_TIME_CODE_ALREADY_REDEEMED,
                            "The verification challenge was already redeemed",
                            null);
                } else {
                    listener.onError(
                            CardPinActionError.UNKNOWN_ERROR,
                            "The call to update the PIN failed, possibly an error " +
                                    "with the verification. Please check the exception.",
                            e);
                }
            } catch (APIConnectionException |
                    APIException |
                    AuthenticationException |
                    CardException e) {
                listener.onError(
                        CardPinActionError.UNKNOWN_ERROR,
                        "An error occurred retrieving the PIN " +
                                "please check the exception",
                        e);
            }
        }
    }

    @Override
    public void onKeyError(@Nullable String operationId,
                           int errorCode,
                           @NonNull String errorMessage) {

        final IssuingCardPinUpdateListener updateListener = mUpdateListeners.remove(operationId);
        final IssuingCardPinRetrievalListener retrievalListener =
                mRetrievalListeners.remove(operationId);
        if (retrievalListener != null) {
            retrievalListener.onError(
                    CardPinActionError.EPHEMERAL_KEY_ERROR,
                    errorMessage,
                    null);
        } else if (updateListener != null) {
            updateListener.onError(
                    CardPinActionError.EPHEMERAL_KEY_ERROR,
                    errorMessage,
                    null);
        }
    }

    public enum CardPinActionError {
        UNKNOWN_ERROR,
        EPHEMERAL_KEY_ERROR,
        ONE_TIME_CODE_INCORRECT,
        ONE_TIME_CODE_EXPIRED,
        ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
        ONE_TIME_CODE_ALREADY_REDEEMED,
    }

    public interface IssuingCardPinRetrievalListener {
        void onIssuingCardPinRetrieved(@NonNull String pin);

        void onError(
                @NonNull CardPinActionError errorCode,
                @Nullable String errorMessage,
                @Nullable Throwable exception);
    }

    public interface IssuingCardPinUpdateListener {
        void onIssuingCardPinUpdated();

        void onError(
                @NonNull CardPinActionError errorCode,
                @Nullable String errorMessage,
                @Nullable Throwable exception);
    }

}

File: stripe/src/main/java/com/stripe/android/IssuingCardPinService.kt
package com.stripe.android

import android.content.Context
import android.util.Log
import androidx.annotation.VisibleForTesting
import com.stripe.android.EphemeralKeyManager.KeyManagerListener
import com.stripe.android.Stripe.Companion.appInfo
import com.stripe.android.exception.InvalidRequestException
import com.stripe.android.networking.StripeApiRepository
import com.stripe.android.networking.StripeRepository
import kotlinx.coroutines.runBlocking

/**
 * Methods for retrieval / update of a Stripe Issuing card
 */
class IssuingCardPinService @VisibleForTesting internal constructor(
    keyProvider: EphemeralKeyProvider,
    private val stripeRepository: StripeRepository,
    private val operationIdFactory: OperationIdFactory = StripeOperationIdFactory()
) {
    private val retrievalListeners = mutableMapOf<String, IssuingCardPinRetrievalListener>()
    private val updateListeners = mutableMapOf<String, IssuingCardPinUpdateListener>()

    private val ephemeralKeyManager = EphemeralKeyManager(
        keyProvider,
        object : KeyManagerListener {
            override fun onKeyUpdate(
                ephemeralKey: EphemeralKey,
                operation: EphemeralOperation
            ) {
                when (operation) {
                    is EphemeralOperation.Issuing.RetrievePin -> {
                        retrievalListeners.remove(operation.id)?.let { listener ->
                            fireRetrievePinRequest(
                                ephemeralKey,
                                operation,
                                listener
                            )
                        } ?: logMissingListener()
                    }
                    is EphemeralOperation.Issuing.UpdatePin ->
                        updateListeners.remove(operation.id)?.let { listener ->
                            fireUpdatePinRequest(
                                ephemeralKey,
                                operation,
                                listener
                            )
                        } ?: logMissingListener()
                }
            }

            override fun onKeyError(
                operationId: String,
                errorCode: Int,
                errorMessage: String
            ) {
                val updateListener = updateListeners.remove(operationId)
                val retrievalListener = retrievalListeners.remove(operationId)
                retrievalListener?.onError(
                    CardPinActionError.EPHEMERAL_KEY_ERROR,
                    errorMessage,
                    null
                ) ?: updateListener?.onError(
                    CardPinActionError.EPHEMERAL_KEY_ERROR,
                    errorMessage,
                    null
                )
            }
        },
        operationIdFactory,
        true
    )

    /**
     * Retrieves a PIN for a given card
     *
     * @param cardId the ID of the card (looks like ic_abcdef1234)
     * @param verificationId the ID of the verification that was sent to the cardholder
     * (typically server-side, through /v1/issuing/verifications)
     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
     * @param listener a listener for either the PIN, or any error that can occur
     */
    fun retrievePin(
        cardId: String,
        verificationId: String,
        userOneTimeCode: String,
        listener: IssuingCardPinRetrievalListener
    ) {
        val operationId = operationIdFactory.create()
        retrievalListeners[operationId] = listener
        ephemeralKeyManager.retrieveEphemeralKey(
            EphemeralOperation.Issuing.RetrievePin(
                cardId = cardId,
                verificationId = verificationId,
                userOneTimeCode = userOneTimeCode,
                id = operationId
            )
        )
    }

    /**
     * Retrieves a PIN for a given card
     *
     * @param cardId the ID of the card (looks like ic_abcdef1234)
     * @param newPin the new desired PIN
     * @param verificationId the ID of the verification that was sent to the cardholder
     * (typically server-side, through /v1/issuing/verifications)
     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
     * @param listener a listener for either the PIN, or any error that can occur
     */
    fun updatePin(
        cardId: String,
        newPin: String,
        verificationId: String,
        userOneTimeCode: String,
        listener: IssuingCardPinUpdateListener
    ) {
        val operationId = operationIdFactory.create()
        updateListeners[operationId] = listener
        ephemeralKeyManager.retrieveEphemeralKey(
            EphemeralOperation.Issuing.UpdatePin(
                cardId = cardId,
                newPin = newPin,
                verificationId = verificationId,
                userOneTimeCode = userOneTimeCode,
                id = operationId
            )
        )
    }

    private fun fireRetrievePinRequest(
        ephemeralKey: EphemeralKey,
        operation: EphemeralOperation.Issuing.RetrievePin,
        listener: IssuingCardPinRetrievalListener
    ) {
        runCatching {
            runBlocking {
                stripeRepository.retrieveIssuingCardPin(
                    operation.cardId,
                    operation.verificationId,
                    operation.userOneTimeCode,
                    ephemeralKey.secret
                )
            }
        }.fold(
            onSuccess = listener::onIssuingCardPinRetrieved,
            onFailure = {
                onRetrievePinError(it, listener)
            }
        )
    }

    private fun onRetrievePinError(
        throwable: Throwable,
        listener: IssuingCardPinRetrievalListener
    ) {
        when (throwable) {
            is InvalidRequestException -> {
                when (throwable.stripeError?.code) {
                    "expired" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_EXPIRED,
                            "The one-time code has expired",
                            null
                        )
                    }
                    "incorrect_code" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_INCORRECT,
                            "The one-time code was incorrect.",
                            null
                        )
                    }
                    "too_many_attempts" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
                            "The verification challenge was attempted too many times.",
                            null
                        )
                    }
                    "already_redeemed" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_ALREADY_REDEEMED,
                            "The verification challenge was already redeemed.",
                            null
                        )
                    }
                    else -> {
                        listener.onError(
                            CardPinActionError.UNKNOWN_ERROR,
                            "The call to retrieve the PIN failed, possibly an error with the verification.",
                            throwable
                        )
                    }
                }
            }
            else -> {
                listener.onError(
                    CardPinActionError.UNKNOWN_ERROR,
                    "An error occurred while retrieving the PIN.",
                    throwable
                )
            }
        }
    }

    private fun fireUpdatePinRequest(
        ephemeralKey: EphemeralKey,
        operation: EphemeralOperation.Issuing.UpdatePin,
        listener: IssuingCardPinUpdateListener
    ) {
        runCatching {
            runBlocking {
                stripeRepository.updateIssuingCardPin(
                    operation.cardId,
                    operation.newPin,
                    operation.verificationId,
                    operation.userOneTimeCode,
                    ephemeralKey.secret
                )
            }
        }.fold(
            onSuccess = {
                listener.onIssuingCardPinUpdated()
            },
            onFailure = {
                onUpdatePinError(it, listener)
            }
        )
    }

    private fun onUpdatePinError(throwable: Throwable, listener: IssuingCardPinUpdateListener) {
        when (throwable) {
            is InvalidRequestException -> {
                when (throwable.stripeError?.code) {
                    "expired" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_EXPIRED,
                            "The one-time code has expired.",
                            null
                        )
                    }
                    "incorrect_code" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_INCORRECT,
                            "The one-time code was incorrect.",
                            null
                        )
                    }
                    "too_many_attempts" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
                            "The verification challenge was attempted too many times.",
                            null
                        )
                    }
                    "already_redeemed" -> {
                        listener.onError(
                            CardPinActionError.ONE_TIME_CODE_ALREADY_REDEEMED,
                            "The verification challenge was already redeemed.",
                            null
                        )
                    }
                    else -> {
                        listener.onError(
                            CardPinActionError.UNKNOWN_ERROR,
                            "The call to update the PIN failed, possibly an error with the verification.",
                            throwable
                        )
                    }
                }
            }
            else -> {
                listener.onError(
                    CardPinActionError.UNKNOWN_ERROR,
                    "An error occurred while updating the PIN.",
                    throwable
                )
            }
        }
    }

    private fun logMissingListener() {
        Log.e(TAG, "${this::class.java.name} was called without a listener")
    }

    enum class CardPinActionError {
        UNKNOWN_ERROR, EPHEMERAL_KEY_ERROR, ONE_TIME_CODE_INCORRECT,
        ONE_TIME_CODE_EXPIRED, ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
        ONE_TIME_CODE_ALREADY_REDEEMED
    }

    interface IssuingCardPinRetrievalListener : Listener {
        fun onIssuingCardPinRetrieved(pin: String)
    }

    interface IssuingCardPinUpdateListener : Listener {
        fun onIssuingCardPinUpdated()
    }

    interface Listener {
        fun onError(
            errorCode: CardPinActionError,
            errorMessage: String?,
            exception: Throwable?
        )
    }

    companion object {
        private val TAG = IssuingCardPinService::class.java.name

        /**
         * Create a IssuingCardPinService with the provided [EphemeralKeyProvider].
         *
         * @param keyProvider an [EphemeralKeyProvider] used to obtain an [EphemeralKey]
         */
        @JvmStatic
        fun create(
            context: Context,
            keyProvider: EphemeralKeyProvider
        ): IssuingCardPinService {
            return create(
                context,
                PaymentConfiguration.getInstance(context).publishableKey,
                keyProvider
            )
        }

        /**
         * Create a [IssuingCardPinService] with the provided [EphemeralKeyProvider].
         *
         * @param publishableKey an API publishable key
         * @param keyProvider an [EphemeralKeyProvider] used to obtain an [EphemeralKey]
         */
        @JvmStatic
        fun create(
            context: Context,
            publishableKey: String,
            keyProvider: EphemeralKeyProvider
        ): IssuingCardPinService {
            return IssuingCardPinService(
                keyProvider,
                StripeApiRepository(context, publishableKey, appInfo),
                StripeOperationIdFactory()
            )
        }
    }
}
